packages/rewriter/lib/js-rules.ts

```diff
@@ -0,0 +1,159 @@
+import { RewriteNodeFn } from './js'
+import { MemberExpressionKind } from 'ast-types/gen/kinds'
+
+// need to inject a ternary to determine if `prop === window.prop` (has not been redefined in a closure), now we're having fun
+// const closureDetectionTern = (prop) => {
+//   return `(${prop} === window['${prop}'] ? ${match('window', prop)} : ${prop})`
+// }
+
+/**
+ * Given an ESTree node and the corresponding source code, either:
+ *  1. returns a string that replaces the node in `js`, or
+ *  2. returns nothing to not modify the node at all
+ *
+ * @see https://runkit.com/flotwig/esprima-doodle for experiments in
+ *  converting common JS constructs to ESTree nodes
+ *
+ * @param js Full JS source that `node` comes from
+ * @param node ESTree node with range metadata
+ */
+export const rewriteJsFnsCb: RewriteNodeFn = (_js, n) => {
+  const b = n.builders
+
+  // use `globalThis` instead of `window`, `self`... to lower chances of scope conflict
+  // users can technically override even this, but it would be very rude
+  // "[globalThis] provides a way for polyfills/shims, build tools, and portable code to have a reliable non-eval means to access the global..."
+  // @see https://github.com/tc39/proposal-global/blob/master/NAMING.md
+  const globalIdentifier = b.identifier('globalThis')
+
+  function match (accessedObject, prop: string, maybeVal?: any) {
+    const args = [
+      // window
+      globalIdentifier,
+      // accessedObject
+      accessedObject,
+      // 'prop'
+      b.stringLiteral(prop),
+    ]
+
+    if (maybeVal) {
+      // maybeVal is a Node
+      args.push(maybeVal)
+    }
+
+    return b.callExpression(
+      b.memberExpression(
+        b.memberExpression(
+          b.memberExpression(
+            globalIdentifier,
+            b.identifier('top'),
+          ),
+          b.identifier('Cypress'),
+        ),
+        b.identifier('resolveWindowReference'),
+      ),
+      args,
+    )
+  }
+
+  // (PROP === window['PROP'] ? MATCH : PROP)
+  function closureDetectionTern (prop) {
+    return b.conditionalExpression(
+      b.binaryExpression(
+        '===',
+        b.identifier(prop),
+        b.memberExpression(
+          globalIdentifier,
+          b.stringLiteral(prop),
+          true,
+        ),
+      ),
+      match(globalIdentifier, prop),
+      b.identifier(prop),
+    )
+  }
+
+  function getReplaceablePropOfMemberExpression (node) {
+    // something.(top|parent)
+    if (node.property.type === 'Identifier' && ['parent', 'top', 'location', 'frames'].includes(node.property.name)) {
+      return node.property.name
+    }
+
+    // something['(top|parent)']
+    if (node.property.type === 'Literal' && ['parent', 'top', 'location', 'frames'].includes(String(node.property.value))) {
+      return String(node.property.value)
+    }
+  }
+
+  return {
+    visitMemberExpression (path) {
```

Comment:
probably worth adding comments above these visitors that describe the intended transformations